mcs/class/Mono.Posix/Test/Mono.Unix.Native/SocketTest.cs

```diff
@@ -0,0 +1,215 @@
+//
+// socket-related test cases
+//
+// Authors:
+//  Steffen Kiess (s-kiess@web.de)
+//
+// Copyright (C) 2015 Steffen Kiess
+//
+
+using System;
+using System.IO;
+using System.Net;
+using System.Net.Sockets;
+using System.Runtime.InteropServices;
+
+using Mono.Unix;
+using Mono.Unix.Native;
+
+using NUnit.Framework;
+
+namespace MonoTests.Mono.Unix.Native
+{
+	[TestFixture, Category ("NotDotNet")]
+	public class SocketTest {
+
+		string TempFolder;
+
+		[SetUp]
+		public void SetUp ()
+		{
+			TempFolder = Path.Combine (Path.GetTempPath (), this.GetType ().FullName);
+
+			if (Directory.Exists (TempFolder))
+				Directory.Delete (TempFolder, true);
+
+			Directory.CreateDirectory (TempFolder);
+		}
+
+		[TearDown]
+		public void TearDown()
+		{
+			if (Directory.Exists (TempFolder))
+				Directory.Delete (TempFolder, true);
+		}
+
+		// Set a timeout on all sockets to make sure that if a test fails it
+		// won't cause the program to hang
+		void SetTimeout (int socket)
+		{
+			Timeval timeout = new Timeval { tv_sec = 0, tv_usec = 500000 };
+			if (Syscall.setsockopt (socket, UnixSocketProtocol.SOL_SOCKET, UnixSocketOptionName.SO_RCVTIMEO, timeout) < 0)
+				UnixMarshal.ThrowExceptionForLastError ();
+			if (Syscall.setsockopt (socket, UnixSocketProtocol.SOL_SOCKET, UnixSocketOptionName.SO_SNDTIMEO, timeout) < 0)
+				UnixMarshal.ThrowExceptionForLastError ();
+		}
+
+		void WithSocketPair (Action<int, int> f)
+		{
+			var sockets = new int[2];
+			if (Syscall.socketpair (UnixAddressFamily.AF_UNIX, UnixSocketType.SOCK_STREAM, 0, sockets) < 0)
+				UnixMarshal.ThrowExceptionForLastError ();
+			try {
+				SetTimeout (sockets[0]);
+				SetTimeout (sockets[1]);
+
+				f (sockets[0], sockets[1]);
+			} finally {
+				if (Syscall.close (sockets[0]) < 0)
+					UnixMarshal.ThrowExceptionForLastError ();
+				if (Syscall.close (sockets[1]) < 0)
+					UnixMarshal.ThrowExceptionForLastError ();
+			}
+		}
+
+		void WithSockets (UnixAddressFamily af, UnixSocketType type, UnixSocketProtocol protocol, Action<int, int> f)
+		{
+			int so1, so2;
+			if ((so1 = Syscall.socket (af, type, protocol)) < 0)
+				UnixMarshal.ThrowExceptionForLastError ();
+			try {
+				if ((so2 = Syscall.socket (af, type, protocol)) < 0)
+					UnixMarshal.ThrowExceptionForLastError ();
+				try {
```

Comment:
Because WithSocketPair() has only a single call to socketpair() which either succeeds (producing two fds) or fails (producing no fd) so there is no case where only a single fd has to be closed.